include/RAJA/pattern/kernel/Lambda.hpp

```diff
@@ -60,32 +128,163 @@ struct Lambda : internal::Statement<camp::nil> {
+    ((camp::at_v<segment_types, OffsetIdx>)(camp::get<OffsetIdx>(data.segment_tuple).begin()[camp::get<OffsetIdx>(data.offset_tuple)])...,
+     camp::get<ParamIdx>(data.param_tuple)...);
+}
+
+
+template <camp::idx_t LoopIndex, typename Types, typename Data>
+RAJA_INLINE RAJA_HOST_DEVICE void invoke_lambda(Data &&data)
+{
+  using Data_t = camp::decay<Data>;
+  using offset_tuple_t = typename Data_t::offset_tuple_t;
+  using param_tuple_t = typename Data_t::param_tuple_t;
+
+  invoke_lambda_expanded<LoopIndex, Types>(
+      camp::make_idx_seq_t<camp::tuple_size<offset_tuple_t>::value>{},
+      camp::make_idx_seq_t<camp::tuple_size<param_tuple_t>::value>{},
+      std::forward<Data>(data));
+}
+
+
+template <camp::idx_t LambdaIndex, typename Types>
+struct StatementExecutor<statement::Lambda<LambdaIndex>, Types> {
 
   template <typename Data>
-  static RAJA_INLINE void exec(Data &&data)
+  static RAJA_INLINE RAJA_HOST_DEVICE void exec(Data &&data)
+  {
+    invoke_lambda<LambdaIndex, Types>(std::forward<Data>(data));
+  }
+};
+
+
+
+
+//Extracts arguments from segments, and parameters
+template<typename Types, typename T>
+struct LambdaArgExtractor;
+
+template<typename Types, camp::idx_t id>
+struct LambdaArgExtractor<Types, LambdaArg<lambda_arg_offset_t, id>>
+{
+
+  // extract offset value type from LoopTypes
+  using type = camp::at_v<typename Types::offset_types_t, id>;
+
+  static_assert(!std::is_same<type, void>::value,
+      "Offset not assigned, but used in Lambda with Offsets<> argument");
+
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  constexpr
+  static type extract_arg(Data &&data)
+  {
+    return type(camp::get<id>(data.offset_tuple));
+  }
+
+};
+
+template<typename Types, camp::idx_t id>
+struct LambdaArgExtractor<Types, LambdaArg<lambda_arg_seg_t, id>>
+{
+
+  // extract segment value type from LoopTypes
+  using type = camp::at_v<typename Types::segment_types_t, id>;
+
+  static_assert(!std::is_same<type, void>::value,
+      "Segment not assigned, but used in Lambda with Segs<> argument");
+
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  constexpr
+  static type extract_arg(Data &&data)
   {
-    invoke_lambda<LambdaIndex>(std::forward<Data>(data));
+    return type(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)]);
   }
+
 };
 
+template<typename Types, camp::idx_t id>
+struct LambdaArgExtractor<Types, LambdaArg<lambda_arg_param_t, id>>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  constexpr
+  static auto extract_arg(Data &&data)->
+    typename std::add_lvalue_reference<camp::tuple_element_t<id,typename camp::decay<Data>::param_tuple_t>>::type
+  {
+    return camp::get<id>(data.param_tuple);
+  }
+};
+
+
+template<typename Types, typename T, camp::idx_t value>
+struct LambdaArgExtractor<Types, LambdaArg<lambda_arg_value_t<T>, value>>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
```

Comment:
@ajkunen pardon my ignorance, but does an inline directive add anything to a constexpr method. I would think that a compiler would simply replace a call to this with the code in the body.

Reply:
unfortunately, constexpr doesn't FORCE the compiler to evaluate it at compile time.  The compiler has the option to actually generate a function/symbol.  I've run into this on a couple of compilers.  The RAJA_INLINE help "convince" the compiler to "do the right thing"

Reply:
Really?!? Good to know. Thanks!